egs/gigaspeech/ASR/conformer_ctc/asr_datamodule.py

```diff
@@ -0,0 +1,382 @@
+            "PyTorch DataLoaders from Lhotse CutSet's -- they control the "
+            "effective batch sizes, sampling strategies, applied data "
+            "augmentations, etc.",
+        )
+        group.add_argument(
+            "--manifest-dir",
+            type=Path,
+            default=Path("data/fbank"),
+            help="Path to directory with train/valid/test cuts.",
+        )
+        group.add_argument(
+            "--max-duration",
+            type=int,
+            default=200.0,
+            help="Maximum pooled recordings duration (seconds) in a "
+            "single batch. You can reduce it if it causes CUDA OOM.",
+        )
+        group.add_argument(
+            "--bucketing-sampler",
+            type=str2bool,
+            default=False,
+            help="When enabled, the batches will come from buckets of "
+            "similar duration (saves padding frames).",
+        )
+        group.add_argument(
+            "--num-buckets",
+            type=int,
+            default=30,
+            help="The number of buckets for the BucketingSampler"
+            "(you might want to increase it for larger datasets).",
+        )
+        group.add_argument(
+            "--concatenate-cuts",
+            type=str2bool,
+            default=False,
+            help="When enabled, utterances (cuts) will be concatenated "
+            "to minimize the amount of padding.",
+        )
+        group.add_argument(
+            "--duration-factor",
+            type=float,
+            default=1.0,
+            help="Determines the maximum duration of a concatenated cut "
+            "relative to the duration of the longest cut in a batch.",
+        )
+        group.add_argument(
+            "--gap",
+            type=float,
+            default=1.0,
+            help="The amount of padding (in seconds) inserted between "
+            "concatenated cuts. This padding is filled with noise when "
+            "noise augmentation is used.",
+        )
+        group.add_argument(
+            "--on-the-fly-feats",
+            type=str2bool,
+            default=False,
+            help="When enabled, use on-the-fly cut mixing and feature "
+            "extraction. Will drop existing precomputed feature manifests "
+            "if available.",
+        )
+        group.add_argument(
+            "--shuffle",
+            type=str2bool,
+            default=True,
+            help="When enabled (=default), the examples will be "
+            "shuffled for each epoch.",
+        )
+        group.add_argument(
+            "--return-cuts",
+            type=str2bool,
+            default=True,
+            help="When enabled, each batch will have the "
+            "field: batch['supervisions']['cut'] with the cuts that "
+            "were used to construct it.",
+        )
+
+        group.add_argument(
+            "--num-workers",
+            type=int,
+            default=20,
```

Comment:
Oh, this value is too large, I think, it may cause CPU OOM.
We are using `2` in LibriSpeech. It was 4, which caused CPU OOM when there are limited CPU RAM.

Reply:
fixed

Reply:
I think he’s using lazy manifests in which case the RAM usage is greatly reduced with more workers. But it won’t play super well with bucketing (I am planning a “on the fly” bucketing mechanism in the future)

Reply:
@wgb14 how many CPU RAM it takes to run GigaSpeech recipe with your asr_datamodule.py setting? 

https://github.com/k2-fsa/icefall/pull/120/files/bea78f609445a407db2377304818da550268d79c#diff-13aabaed6dbf8573df7a9b01e6c22c716b4cff37948c98055b915a010404bd63

Reply:
FYI @wgb14 @pingfengluo you can now improve the RAM usage (and I think by a lot) by doing the following:

1) open manifests lazily (i.e., `CutSet.from_jsonl_lazy(path)` or `lhotse.load_manifest_lazy(path)`, both are equivalent)

2) replace `BucketingSampler` with `DynamicBucketingSampler` (you might need to remove/change some arguments, but they are mostly the same)

Reply:
@pingfengluo updated my `asr_datamodule.py` setting https://github.com/k2-fsa/icefall/pull/120/commits/6e5b189fc5509c76bc2d989710ee85e78ad6603b
I didn't get CPU OOM with DynamicBucketingSampler and lazy-loading training set.

Reply:
@pzelasko with DynamicBucketingSampler,  there is still OOM on WenetSpeech dataset  training on a 256G RAM machine, my asr_datamodule.py is https://github.com/pingfengluo/icefall/blob/wenetspeech/egs/wenetspeech/ASR/tdnn_lstm_ctc/asr_datamodule.py .
should i checkout this PR https://github.com/k2-fsa/icefall/pull/120#discussion_r776461928 ?

Reply:
@pingfengluo can you use on the fly feature extraction and see if the issue persists?

Reply:
> @pingfengluo can you use on the fly feature extraction and see if the issue persists?

i have tried on-the-fly feature extraction, but there still was OOM.

Reply:
You probably haven’t opened the CutSet lazily (`CutSet.from_jsonl_lazy(path)`)

Reply:
> You probably haven’t opened the CutSet lazily (`CutSet.from_jsonl_lazy(path)`)

I make sure  than using  `CutSet.from_jsonl_lazy(path)`  ,  asr_datamodule.py and log as  https://github.com/k2-fsa/icefall/pull/167/files#diff-8951e342449f8076c154f4382e8c56579448dd8bc41e021a1ddf9794e115a5c8
![image](https://user-images.githubusercontent.com/2265062/147866649-7410378e-3be3-4a3d-bad3-4b9cedcb9ca7.png)

Reply:
Does memory usage seem to increase gradually?
I wonder whether it dies while processing an exceptionally long utterance?
Don't lose hope!  I'm sure we are getting closer to a solution.

Reply:
... and please clarify every time you mention OOM whether it is CUDA or CPU, because the comments on this PR mention both types.

Reply:
Is there any specific function call that makes it go OOM?

As a test, can you comment out the code for forward and backward pass, just to let it sample batches from the data pipeline, and see if you’re still getting OOM? It will help determine if it’s data pipeline related or not. I didn’t see anything obviously wrong in the data module.

Reply:
> Does memory usage seem to increase gradually? 
Yes. The CPU RAM  increased gradually until using up 256GB.

commands on WenetSpeech recipe are 
```
cd icefall/egs/wenetspeech/ASR && ./transducer_stateless/train.py \
  --bucketing-sampler True \
  --world-size 8 \
  --lang-dir data/lang_char \
  --num-epochs 40 \
  --start-epoch 0 \
  --exp-dir transducer_stateless/exp_char \
  --max-duration 160 \
  --lr-factor 3
```
code to test  is in this PR https://github.com/k2-fsa/icefall/pull/167

> I wonder whether it dies while processing an exceptionally long utterance? 
I will try to remove long utterance.

> I'm sure we are getting closer to a solution.
Great!

Reply:
> As a test, can you comment out the code for forward and backward pass, just to let it sample batches from the data pipeline, and see if you’re still getting OOM? 

Ok,  i will test it. thanks.